bukhr/i18n-recursive-lookup
# Feature/optional disable cache compiled values

# English
## Problem description
It is necessary to fork the `I18n-recursive-lookup` gem because, upon analyzing its flow, we realized that it caches the values of the translations made on the first request. In consequence, the tenants save the values of several translations that were queried when the application started. This poses a problem for translations that change based on user interactions, for instance, because the translation will always return the first cached value regardless of what the user does.

## What did we do to solve the problem?
The problem was found in the call to the I18n function `store_translations` which is made inside the `cache_compiled_result` function of the recursive gem. Additionally, the recursive hashes were cached due to an in-place replacement that occurs within the `deep_compile` function. To make the change less invasive, we left the behavior of disabled cache as optional and by default it uses cache, for which the function `enabled_interpolation_cache?` was added. To disable the cache, simply add `I18n.backend.enabled_interpolation_cache = false`.

Furthermore, a new class of tests was added to test this new disabled cache behavior, and a test that was broken in the original gem repository was fixed.

## How to test?
1. Download the repository and install dependencies
2. Run tests with `rake test test/backend/recursive_lookup_test.rb` and verify all tests pass.

# Español
## Issues relacionados
Contexto: [Slack](https://bukhr.slack.com/archives/C03PSQ4RBCZ/p1676396826687299)

## Descripción del problema
Se requiere hacer un fork de la gema `I18n-recursive-lookup` ya que al analizar su flujo nos dimos cuenta que cachea los valores de las traducciones que hace al primer request, lo cual provoca que el tenant obtenga y se quede con los valores de ciertas traducciones que consulta al iniciarse la aplicación. En particular, esto hubiera sido un gran problema para cuando se implemente el [campo de cambio de palabra "empleado"](https://buk.atlassian.net/browse/HCM-2188?atlOrigin=eyJpIjoiZDE1OGY4ZWRmZjdhNDMwZjliZGQ0NTY1OGE3ODM2NWIiLCJwIjoiaiJ9): el cambio no se observaría porque la traducción iría al primer valor cacheado y no se actualizaría con la nueva palabra ingresada por el usuario. 

## Qué se hizo para resolver el problema
Se vio que el problema estaba en el llamado a la función de I18n `store_translations` que se hace dentro de la función `cache_compiled_result` de la gema recursiva. Además, los hash recursivos quedaban cacheados debido a un reemplazo in-place que se realiza dentro de la función `deep_compile`. Para que el cambio no fuera tan invasivo, dejamos el comportamiento de no cachear las traducciones como opcional y que por defecto utilice caché, para lo cual se agregó la función `enabled_interpolation_cache?`. Para cambiar el comportamiento para que no ocupe caché basta con agregar `I18n.backend.enabled_interpolation_cache = false`.

Además, se agregó una nueva clase de tests para probar este nuevo comportamiento sin caché y se modificó un test que venía roto desde el repo de la gema original.

## Cómo probar
1. Descargar el repositorio e instalar dependencias
2. Correr los tests con `rake test test/backend/recursive_lookup_test.rb` y verificar que pasen todos.
